File: 06-subprogramas/codes/referencia.c
#include <stdio.h>

void incrementar(int *numeroPtr) { 
    printf("Dentro da funcao: O valor inicial apontado pelo ponteiro de numero eh %d\n", *numeroPtr);
    *numeroPtr = *numeroPtr + 10; // Altera o valor no ENDERECO de memoria
    printf("Dentro da funcao: O valor apos a alteracao apontado pelo ponteiro de numero eh %d\n", *numeroPtr);
}

int main() {
    int num = 5;
    printf("Antes da chamada da funcao: O valor do numero eh %d\n", num);

    incrementar(&num); // Passa o ENDERECO de memoria de num

    printf("Depois da chamada da funcao: O valor do numero eh %d\n", num); // num agora será 15
    return 0;
}

File: 06-subprogramas/codes/referencia.java
class Pessoa {
    String nome;
    int idade;

    public Pessoa(String nome, int idade) {
        this.nome = nome;
        this.idade = idade;
    }

    @Override
    public String toString() {
        return "Pessoa [nome: " + nome + ", idade: " + idade + "]";
    }
}

public class referencia {
    public static void modificarPessoa(Pessoa pessoa) {
        System.out.println("Dentro do metodo (antes da modificacao): " + pessoa);

        pessoa.idade = 20;
        pessoa.nome = "Maria";
        System.out.println("Dentro do metodo (depois da modificacao de estado): " + pessoa);
    }

    public static void main(String[] args) {
        Pessoa pessoa = new Pessoa("Victor", 23);
        System.out.println("Antes da chamada do metodo: " + pessoa);

        modificarPessoa(pessoa); // Passa a REFERENCIA 

        System.out.println("Depois da chamada do metodo: " + pessoa);
    }
}